engines/wintermute/base/gfx/opengl/base_render_opengl3d_shader.cpp

```diff
@@ -0,0 +1,607 @@
+	glGetTexLevelParameteriv(GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &texHeight);
+
+	float texLeft = (float)rect.left / (float)texWidth;
+	float texTop = (float)rect.top / (float)texHeight;
+	float texRight = (float)rect.right / (float)texWidth;
+	float texBottom = (float)rect.bottom / (float)texHeight;
+
+	float offset = _viewportRect.height() / 2.0f;
+	float correctedYPos = (pos.y - offset) * -1.0f + offset;
+
+	// to be implemented
+	if (mirrorX) {
+		warning("BaseRenderOpenGL3DShader::SpriteEx x mirroring is not yet implemented");
+	}
+
+	if (mirrorY) {
+		warning("BaseRenderOpenGL3DShader::SpriteEx y mirroring is not yet implemented");
+	}
+
+	SpriteVertexShader vertices[4] = {};
+
+	// texture coords
+	vertices[0].u = texLeft;
+	vertices[0].v = texTop;
+
+	vertices[1].u = texLeft;
+	vertices[1].v = texBottom;
+
+	vertices[2].u = texRight;
+	vertices[2].v = texTop;
+
+	vertices[3].u = texRight;
+	vertices[3].v = texBottom;
+
+	// position coords
+	vertices[0].x = pos.x - 0.5f;
+	vertices[0].y = correctedYPos - 0.5f;
+
+	vertices[1].x = pos.x - 0.5f;
+	vertices[1].y = correctedYPos - height - 0.5f;
+
+	vertices[2].x = pos.x + width - 0.5f;
+	vertices[2].y = correctedYPos - 0.5f;
+
+	vertices[3].x = pos.x + width - 0.5f;
+	vertices[3].y = correctedYPos - height - 0.5;
+
+	// not exactly sure about the color format, but this seems to work
+	byte a = RGBCOLGetA(color);
+	byte r = RGBCOLGetR(color);
+	byte g = RGBCOLGetG(color);
+	byte b = RGBCOLGetB(color);
+
+	for (int i = 0; i < 4; ++i) {
+		vertices[i].r = r / 255.0f;
+		vertices[i].g = g / 255.0f;
+		vertices[i].b = b / 255.0f;
+		vertices[i].a = a / 255.0f;
+	}
+
+	// transform vertices here if necessary, add offset
+
+	_spriteShader->use();
+	_spriteShader->setUniform("alphaTest", !alphaDisable);
+	_spriteShader->setUniform("projMatrix", _projectionMatrix2d);
+
+	glBindBuffer(GL_ARRAY_BUFFER, _spriteVBO);
+	glBufferSubData(GL_ARRAY_BUFFER, 0, 4 * sizeof(SpriteVertexShader), vertices);
+
+	setSpriteBlendMode(blendMode);
+
+	glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+
+	return true;
+}
+
+void BaseRenderOpenGL3DShader::renderSceneGeometry(BaseArray<AdWalkplane *> &planes, BaseArray<AdBlock *> &blocks, BaseArray<AdGeneric *> &generics, Camera3D *camera) {
+	// don't render scene geometry, as OpenGL ES 2 has no wireframe rendering and we don't have a shader alternative yet
+}
+
+void BaseRenderOpenGL3DShader::renderShadowGeometry(BaseArray<AdWalkplane *> &planes, BaseArray<AdBlock *> &blocks, BaseArray<AdGeneric *> &generics, Camera3D *camera) {
+	resetModelViewTransform();
+	setup3D(camera, true);
+
+	// disable color write
+	glBlendFunc(GL_ZERO, GL_ONE);
+
+	glFrontFace(GL_CCW);
+	glBindTexture(GL_TEXTURE_2D, 0);
+
+	// render walk planes
+	for (uint i = 0; i < planes.size(); i++) {
+		if (planes[i]->_active && planes[i]->_receiveShadows) {
+			planes[i]->_mesh->render();
+			//m_Renderer->m_NumPolygons += _planes[i]->m_Mesh->m_NumFaces;
```

Comment:
what is commented code?

Reply:
Same as above, this might just be removed.

Reply:
Comment removed.